### ISSUE TITLE
BatchNorm kernel?

### ISSUE THREAD

**Issue by lukemetz** [tensorflow/tensorflow#1502]
Hello,
Adding batch norm to a model results in a significant slowdown (using tf.nn.batch_normalize).

I was wondering why a dedicated kernel was removed in favour of a symbolic approach?
(seemingly unrelated commit in question: https://github.com/tensorflow/tensorflow/commit/03869e6e9cfac85d3e0f87ded01f4dd8de10ed67)

Was this a design decisions to not want these "one off" style optimizations/kernels?

Having a faster kernel, (like cudnn #714, or something like what torch does) would be great if its not against tensorflows design. I will look into it if it is not.

Thank you.

**Comment by ajtulloch** [comment]
Yeah, we've been fusing graph ops manually for performance, it'd be great to understand the rationale for going the other way.

**Comment by mrry** [comment]
It looks like two commits were inadvertently squashed. Here's the intended description for the batch norm parts of that commit:

> General batchnorm op, suitable to both global normalization for convolutions and per-depth normalization of fully-connected layers. It also supports arbitrary layouts and orderings of the axes.
> 
> Nothing deep in these essentially 4 lines of code, but note:
> - The ops are ordered in a way that makes the computation much more efficient (thx Sergey). None of the existing wrappers that didn't use the fused kernel did that, and as a result were almost 2x slower according to the benchmark.
> - The benchmark confirms that this implementation is competitive with the handcrafted kernel.
> - I confirmed on inception that the step time or accuracy are not affected.
> - Added documentation and tests showing how the op can be used for arbitrary layouts.
> - The change in API compared with batch_norm_with_global_normalization makes it possible to not have a scale tensor when it's not used, which caused some confusion and unnecessary additional saved variables in the past.

@vincentvanhoucke is working on the related issue https://github.com/tensorflow/tensorflow/issues/1122 and might have more to add.

**Comment by vincentvanhoucke** [comment]
If you're seeing a significant regression from the non-fused kernel, I'd be interested in hearing about it. Note that there were two changes:
- removing the fused kernel, which did not affect performance on any benchmark I could run. In particular, this one:
https://github.com/tensorflow/tensorflow/blob/0249729a26b2cd3bdbbc5880f8d634a56860b0fd/tensorflow/python/ops/batch_norm_benchmark.py
- replacing the two-pass moment accumulation algorithm with a one-pass algorithm. This was a tad slower on GPU, and faster on CPU.
On balance, those two changes gave us a lot more flexibility:
- the one-pass moment accumulation removes a long-lived doubling of the memory footprint of the activations.
- it makes it possible to compute moments over long sequences more efficiently.
- the removal of the fused kernel generalized the computation to many more data shapes and layouts, and will enable for example batch-major and depth-major for convolutions.
We could recreate a fused op that implements the more flexible shapes if it significantly improved performance. It's also possible that a fused kernel for the one-pass moment computation wold be much more efficient. We need to know which is causing regressions first if there are any.

**Comment by lukemetz** [comment]
Thank you all so much for the information. There is no regression, it just seemed slower than I would have guessed from experience in other frameworks. No cross library benchmarks were done.

As for the benchmark, the python version is indeed the fastest on the workloads I care about (CNN).

If I get a free moment I will look into a faster kernel and check performance myself.

**Comment by ppwwyyxx** [comment]
I'm trying out `batch_norm_benchmark.py` with NCHW layout and the performance seems to drop significantly.
I might not test it in the right way: I ran `run_graph()` with `shape=[8,128,128,32],axes=[0,1,2]`(NHWC) and then 
`shape=[8,32,128,128], axes=[0,2,3]`(NCHW), on GPU, with `mode=py`. I saw a 30x performance difference.
Will this be fixed?

**Comment by lukemetz** [comment]
Hmm I actually just ran into this locally on my own cnn models. (Didn't think about that benchmark though.) I did a quick profile with cuda and found `reduce_sum`'s being the big killers in performance (Slowdown's occur in the `batch_normalize` function as well as the `moments` calculation.)

**Comment by vincentvanhoucke** [comment]
@benoitsteiner, a few reports of very poor reduction performance for some layouts (see benchmark above). Can you triage or investigate?

**Comment by vincentvanhoucke** [comment]
@ppwwyyxx @lukemetz if you could confirm which version you're using as well. I know that a lot of work has gone into improving reductions recently, so if you're not using the head of the tree I'd love to get confirmation that this is still an issue.

**Comment by lukemetz** [comment]
@vincentvanhoucke I updated yesterday morning (63409bd23facad471973b110df998782c0e19c06).

**Comment by ppwwyyxx** [comment]
I compiled just now. And use the following main() function in `batch_norm_benchmark.py`:
```python
def main(unused_argv):
  print("Forward convolution (lower layers).")
  shape = [8, 128, 128, 32]
  axes = [0, 1, 2]
  t2 = run_graph("gpu", shape, axes, 10, "py", True, False, 50)
  shape = [8, 32, 128, 128]
  axes = [0, 2, 3]
  t2 = run_graph("gpu", shape, axes, 10, "py", True, False, 50)
```
output:
```
gpu shape:4/3 #layers:10 mode:py scale:True train:False - 0.019783 secs
gpu shape:4/3 #layers:10 mode:py scale:True train:False - 0.365370 secs
```

**Comment by benoitsteiner** [comment]
In the first case (reduction axes = [0, 1, 2]) we have an optimized implementation leverages the gpu fairly well. In the second case (reduction axes = [0, 2, 3]), we use a basic approach that currently leverages few cuda threads to perform the reduction. This is why the second case is much slower. I'll see what can be done to improve the second case.

**Comment by ppwwyyxx** [comment]
I think batch normalization in cudnnv4 uses NCHW layout, and I suppose that will be fast. Not sure if you have plan to integrate those into tensorflow?

**Comment by tatatodd** [comment]
@benoitsteiner Friendly ping to update this bug with the latest status, thanks!

**Comment by jstaker7** [comment]
Using the timeline module I see that I have very slow batch normalization. One example of "moments/sufficient_statistics/mean_ss" took 413.262ms and "gradients/batchnorm/add_1_grad/Sum_1" took 	
413.357ms. I feel like this is a lot slower than I've seen before. As a sanity check, are these along the same order of magnitude you all are seeing in your benchmarks? This is with a Titan X.

**Comment by vincentvanhoucke** [comment]
Is this at the head of the tree? We recently changed:
          mean, variance = nn.moments(inputs, axis, shift=moving_mean)
to:
          shift = math_ops.add(moving_mean, 0)
          mean, variance = nn.moments(inputs, axis, shift=shift)
to fix a race condition, and this slowed things a tiny bit in our benchmarks.
Since you're seeing a slow Sum, I wonder if it's related. What happens if you roll back that change?

**Comment by jstaker7** [comment]
This is using the v0.10 release and not compiled from scratch at the head. Here is a script I built for a quick benchmark around just computing the moments (very non-rigorous, but hopefully helpful):

    import tensorflow as tf
    import datetime

    a = tf.ones([64, 256, 256, 32])
    mean, variance = tf.nn.moments(a, [0, 1, 2])

    with tf.Session() as sess:
        start = datetime.datetime.now()
        result = sess.run([mean, variance])
        end = datetime.datetime.now()

    print(end - start)

Output on CPU: 0:00:01.318907
Output on GPU: 0:00:00.651986

This is basically BN after my first convolution; perhaps these times are actually reasonable? It is a fairly large matrix.

**Comment by jstaker7** [comment]
Just as a followup, in my timeline the Sum op seems really expensive throughout my graph (mostly related to BN and gradients) and is blocking for most of the other computation. Is this expected behavior? 

<img width="768" alt="timeline" src="https://cloud.githubusercontent.com/assets/1364252/18144914/eccdc898-6f7d-11e6-912c-9a3b687001bb.png">

**Comment by vincentvanhoucke** [comment]
That looks over the top to me. Punting so someone who can actually help.

**Comment by mrry** [comment]
If I recall correctly, the 0.10 *release candidate* binary had a performance regression in some of the reduction kernels (on certain GPUs). I believe this is fixed now at HEAD, and will be in the actual *release* which is due any day now.

Can you try with the nightly build and confirm whether this is still a problem?

**Comment by jstaker7** [comment]
Wow. Night and day difference. Using the nightly build I gained about a 3X speedup and all the long summations in the timeline have disappeared. Thank you!

**Comment by mrry** [comment]
That's a relief! Thanks for confirming our suspicions :).

**Comment by ppwwyyxx** [comment]
Batch norm on HCHW layout is still over 10x slower (after half a year) as discussed above. It is mentioned in #1759 that there will be a cudnn integration. Do we have a time frame about that?

**Comment by thuyen** [comment]
It's in the master branch already, named `nn.fused_batch_norm`. Have a look at #4919 and #4899.
